File: ex18.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.HashSet;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        Set<List<Integer>> r = new HashSet<>();
        List<Integer> record = new ArrayList<>();
        int lo, hi;
        int sum;

        
        // Sort the array first
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 3; i++) {
            for (int j = i + 1; j < nums.length - 2; j++) {
                lo = j + 1;
                hi = nums.length - 1;
            
                while (lo < hi) {
                    sum = nums[i] + nums[j] + nums[lo] + nums[hi];
                
                    if (sum == target) {
                        // Add item to list
                        record.add(nums[i]);
                        record.add(nums[j]);
                        record.add(nums[lo]);
                        record.add(nums[hi]);

                        // Add list ot set
                        r.add(new ArrayList<>(record));
                        record.clear();
                        hi--;   lo++;
                    } else if (sum > target)    hi--;
                    else    lo++;
                }
            }
        }

        
        return new ArrayList<>(r);
    }
}

File: code/ex18.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int n = nums.size();    if (n < 4)    return {};
        sort(nums.begin(), nums.end());
        
        vector<vector<int>> res;
        for (int i = 0; i < n - 3; i++) {
            if (i > 0 && nums[i - 1] == nums[i])    continue;
            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1 && nums[j - 1] == nums[j])    continue;
                int a = nums[i], b = nums[j];
                int t = target - a - b;
                if (nums[j + 1] + nums[j + 2] > t)    continue;
                
                // Two pointers for t
                int lo = j + 1, hi = n - 1;
                while (lo < hi) {
                    while (lo > j + 1 && nums[lo] == nums[lo - 1] && lo < hi)    lo++;
                    while (hi < n - 1 && nums[hi] == nums[hi + 1] && lo < hi)    hi--;
                    if (lo >= hi)    continue;
                    int c = nums[lo], d = nums[hi];
                    if (c + d == t)    {res.push_back({a, b, c, d});    lo++;    hi--;}
                    else if (c + d > t)    hi--;
                    else    lo++;
                }   
            }
        }
        
        return res;
    }
};